process/block/shardblock.go

```diff
@@ -1844,3 +1891,57 @@ func (sp *shardProcessor) getBootstrapHeadersInfo(
 
 	return lastSelfNotarizedHeaders
 }
+
+func (sp *shardProcessor) receivedMiniBlock(key []byte) {
+	log.Debug("miniblock key", "key", key)
+	mb, ok := sp.dataPool.MiniBlocks().Get(key)
+	if ok {
+		peerMb, ok := mb.(*block.MiniBlock)
+		if ok && peerMb.Type == block.PeerBlock {
+			log.Debug(fmt.Sprintf("received miniblock of type %s", peerMb.Type))
+
+			sp.startOfEpochDataMutex.Lock()
+
+			defer sp.startOfEpochDataMutex.Unlock()
+			for _, sod := range sp.startOfEpochData {
+				if sod.stillMissing > 0 {
+					for _, mbHash := range sod.peerMiniblocks {
+						if bytes.Equal(mbHash, key) {
+							sod.stillMissing--
+						}
+					}
+				}
+
+				if sod.stillMissing == 0 {
+					if sp.processPeerMiniBlocks(sod) {
```

Comment:
I think this function can be refactored. Getting all the data separately and having a channel which announces that all the data is gathered, only then the processing of each miniblock can be started.

Reply:
fixed